src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsParams.cs

```diff
@@ -0,0 +1,19 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using MediatR;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    // Note: This type should be kept in sync with the one in VisualStudio.LanguageServerClient assembly.
+    internal class RazorDiagnosticsParams : IRequest<RazorDiagnosticsResponse>
+    {
+        public RazorLanguageKind Kind { get; set; }
```

Comment:
So you currently only look at this `Kind` to optimize how to map C#/HTML diagnostics (which I suggested you may want to avoid). Do you envision us wanting to separately handle HTML/C# diagnostics in the future? I ask because currently you're filtering/mapping both HTML/C# diagnostics in one pass vs. handling each language independently.

One thought that comes to mind is how do you handle situations like:
`<input type="value" @bind-value="SomeErroredCSharp">`

I'd imagine you'd want an error for the C# portion but wouldn't want any HTML diagnostic for funky HTML parameters on input (I forget the situations when you get HTML diagnostics but they def exist)

Reply:
> So you currently only look at this Kind to optimize how to map C#/HTML diagnostics (which I suggested you may want to avoid). 

Discussing above, if we do end up keeping the optimization then we'll need this property.

> Do you envision us wanting to separately handle HTML/C# diagnostics in the future?

Possibly. 

Currently we filter by diagnostic codes. This is just a hash-set of string codes we don't like, so it's language agnostic and can be handled in one quick pass.

There is a potential future of needing more advanced semantic filtering. Ex. filtering the unclosed tag diagnostic for Razor components (as we have a `RZ` diagnostic) but not for HTML tags. In that case having a `Kind` may be helpful to understand which filtering "passes" we want to run through. Granted, this can likely be achieved by separating out the diagnostic codes into individual hash sets for HTML & C#.

> One thought that comes to mind is how do you handle situations like:
`<input type="value" @bind-value="SomeErroredCSharp">`
I'd imagine you'd want an error for the C# portion but wouldn't want any HTML diagnostic for funky HTML parameters on input (I forget the situations when you get HTML diagnostics but they def exist)

Not too worried about cases like these as the underlying diagnostic codes which we use for filtering _should_ be different, so we should be able to filter out the unnecessary diagnostics regardless.

Reply:
> Not too worried about cases like these as the underlying diagnostic codes which we use for filtering _should_ be different, so we should be able to filter out the unnecessary diagnostics regardless.

I think in practice you're totally right; however, I can envision there being a lot of complexity/coupling unintentionally introduced by having both C#/HTML run simultaneously. You've convinced me that `Kind` is needed to know what type of diagnostic filtering is needed; however, I think how you currently understand both HTML/C# simultaneously in one pass will cause issues in the future. (this comment is more towards refactoring the workings of the endpoint not the params now)

Reply:
> You've convinced me that Kind is needed to know what type of diagnostic filtering is needed; however, I think how you currently understand both HTML/C# simultaneously in one pass will cause issues in the future. (this comment is more towards refactoring the workings of the endpoint not the params now)

I agree we should have a clear separation of functionality to this regard between HTML & C#. However, is this something you'd like to see addressed in this PR? My hesitation is simply because we don't do any HTML filtering currently, as we don't have HTML diagnostic codes. Would you be open to revisiting a refactor of the filtering logic once we have HTML filtering requirements in place. Concerned we may optimize a design which isn't quite finalized yet.

Reply:
It's definitely something that can be done later. I think the part that gives me hesitation is that we're inspecting HTML diagnostics looking for C# diagnostic codes which doesn't feel great

Reply:
Ah okay, that's easily addressable now https://github.com/dotnet/aspnetcore-tooling/pull/2832/commits/daf922a0f247e8c8fa1acbcdcd7083063ea9b36c.